lib/std/log.zig

```diff
@@ -0,0 +1,131 @@
+const builtin = std.builtin;
+const root = @import("root");
+
+//! std.log is standardized interface for logging which allows for the logging
+//! of libraries using this interface to be formatted and filtered by the
+//! implementer of the root.log function.
+//!
+//! The scope parameter should be used to give context to the logging. For
+//! example, a library called 'libfoo' might use .libfoo as its scope.
+//!
+//! An example root.log might look something like this:
+//!
+//! ```
+//! const std = @import("std");
+//!
+//! pub const log_level: std.log.Level = .warn;
+//!
+//! pub fn log(
+//!     comptime level: std.log.Level,
+//!     comptime scope: @TypeOf(.EnumLiteral),
+//!     comptime format: []const u8,
+//!     args: var,
+//! ) void {
+//!     // Ignore all .info and .warn level logging from sources other than
+//!     // .my_project and .nice_library
+//!     const scope_prefix = "(" ++ switch (scope) {
+//!         .my_project, .nice_library => @tagName(scope),
+//!         else => if (level == .err) @tagName(scope) else return,
+//!     } ++ "): ";
+//!
+//!     const prefix = "[" ++ @tagName(level) ++ "] " ++ scope_prefix;
+//!
+//!     // Print the message to stderr, silently ignoring any errors
+//!     const held = std.debug.getStderrMutex().acquire();
+//!     defer held.release();
+//!     const stderr = std.debug.getStderrStream();
+//!     nosuspend stderr.print(prefix ++ format, args) catch return;
+//! }
+//!
+//! pub fn main() void {
+//!     // Won't be printed as log_level is .warn
+//!     std.log.info(.my_project, "Starting up.\n", .{});
+//!     std.log.err(.nice_library, "Something went very wrong, sorry.\n", .{});
+//!     // Won't be printed as it gets filtered out by our log function
+//!     std.log.warn(.lib_that_logs_too_much, "Added 1 + 1\n", .{});
+//! }
+//! ```
+//! Which produces the following output:
+//! ```
+//! [err] (nice_library): Something went very wrong, sorry.
+//! ```
+
+pub const Level = enum {
+    /// Log everything.
+    info,
+    /// Log only warnings and errors.
+    warn,
+    /// Log only errors.
+    err,
+    /// Log nothing.
+    silent,
+};
+
+/// The default log level is based on build mode. Note that in ReleaseSmall
+/// builds logging is totally disabled (log.fatal still causes a panic though).
+pub const default_level: Level = switch (builtin.mode) {
+    .Debug => .info,
+    .ReleaseSafe => .warn,
+    .ReleaseFast => .err,
+    .ReleaseSmall => .silent,
+};
+
+/// The current log level. This is set to root.log_level if present, otherwise
+/// log.default_level.
+pub const level: Level = if (@hasDecl(root, "log_level"))
+    root.log_level
+else
+    default_level;
+
+fn log(
+    comptime message_level: Level,
+    comptime scope: @Type(.EnumLiteral),
+    comptime format: []const u8,
+    args: var,
+) void {
+    if (@enumToInt(message_level) < @enumToInt(level))
+        return;
+    if (@hasDecl(root, "log")) {
+        root.log(message_level, scope, format, args);
```

Comment:
I'm trying to decide if we should swap the order of these checks. With the current code one must set `root.log_level` to `.info` in order to implement runtime configurable log levels. However, this keeps `root.log` simpler in the case that comptime log level configuration is sufficient.

Reply:
leaning towards keeping this as is. Setting `root.log_level` to `.info` makes sense if you want to handle filtering yourself at runtime.